src/common/tests_dmg_helpers.c

```diff
@@ -173,48 +221,71 @@ run_cmd(const char *command, int *outputfd)
 	if (child_pid == -1) {
 		rc = daos_errno2der(errno);
 		D_ERROR("failed to fork: %s\n", strerror(errno));
-		return rc;
+		D_GOTO(err_tokens, rc);
 	} else if (child_pid == 0) {
+		int exec_err;
+
 		/* child doesn't need the read end of the pipes */
 		close(stdoutfd[0]);
 		close(stderrfd[0]);
 
-		if (dup2(stdoutfd[1], STDOUT_FILENO) == -1)
+		if (dup2(stderrfd[1], STDERR_FILENO) == -1)
 			_exit(errno);
 
-		if (dup2(stderrfd[1], STDERR_FILENO) == -1)
+		if (dup2(stdoutfd[1], STDOUT_FILENO) == -1)
 			_exit(errno);
 
 		close(stdoutfd[1]);
 		close(stderrfd[1]);
 
-		rc = system(command);
-		if (rc == -1)
-			_exit(errno);
-		_exit(rc);
+		rc = execvp(cmd_tokens[0], cmd_tokens);
+
+		/* if we get here, the exec must have failed */
+		exec_err = errno;
+		D_ASSERT(rc == -1);
+		fprintf(stderr, "exec failed: %s\n", strerror(exec_err));
+		_exit(exec_err);
+	}
+
+	D_DEBUG(DB_TEST, "waiting for child process\n");
+	if (waitpid(child_pid, &child_rc, 0) == -1) {
+		rc = daos_errno2der(errno);
+		D_ERROR("wait failed: %s\n", strerror(errno));
+		D_GOTO(err_tokens, rc);
 	}
+	D_DEBUG(DB_TEST, "child process completed\n");
+
+	cmd_free_args(cmd_tokens, tokens_len);
 
 	/* parent doesn't need the write end of the pipes */
 	close(stdoutfd[1]);
 	close(stderrfd[1]);
 
-	D_DEBUG(DB_TEST, "waiting for dmg to finish executing\n");
-	if (wait(&child_rc) == -1) {
-		D_ERROR("wait failed: %s\n", strerror(errno));
-		return daos_errno2der(errno);
-	}
-	D_DEBUG(DB_TEST, "dmg command executed successfully\n");
+	log_stderr_pipe(stderrfd[0]);
 
 	if (child_rc != 0) {
-		D_ERROR("child process failed, rc=%d (%s)\n", child_rc, strerror(child_rc));
+		D_ERROR("child process failed, exit status=%d\n", WEXITSTATUS(child_rc));
+		if (WIFSIGNALED(child_rc))
+			D_ERROR("child signal: %d\n", WTERMSIG(child_rc));
 		close(stdoutfd[0]);
-		log_stderr_pipe(stderrfd[0]);
-		return daos_errno2der(child_rc);
+		return -DER_MISC;
 	}
 
-	close(stderrfd[0]);
```

Comment:
Not sure why this line was removed. Is it important to close (or not close) stderrfd[0] in the parent before returning?